plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/StmtConversionVisitor.kt

```diff
@@ -217,12 +217,15 @@ object StmtConversionVisitor : FirVisitor<ExpEmbedding, StmtConversionContext<Re
             }
             else -> {
                 val retType = implicitInvokeCall.calleeCallableSymbol.resolvedReturnType
-                args.filter { it.type is UnspecifiedFunctionTypeEmbedding }
-                    .forEach {
+                args
+                    .zip(implicitInvokeCall.argumentList.arguments)
+                    .filter { (embedding, _) -> embedding.type is UnspecifiedFunctionTypeEmbedding }
+                    .forEach { (embedding, firArgument) ->
                         val leakFunctionObjectCall = Stmt.Assert(
                             DuplicableFunction.toFuncApp(
-                                listOf(it.pureToViper()),
-                                implicitInvokeCall.source.asPosition
+                                listOf(embedding.pureToViper()),
+                                implicitInvokeCall.source.asPosition,
+                                SourceRole.ParamFunctionLeakageCheck(firArgument.calleeSymbol).asInfo
```

Comment:
I am surprised that you need to be providing a source role here.  This refers to parameters of the callee, but we generally do not care about those; we care about the arguments that we are passing as the caller.  Are you sure that you are reporting the correct function?

Reply:
I think so. I check it double the warning message about the leaking function.

```kotlin
@Suppress("LEAKED_IN_PLACE_LAMBDA")
@OptIn(ExperimentalContracts::class)
fun <!VIPER_TEXT!>function_object_call<!>(f: (() -> Unit) -> Unit, g: () -> Unit) {
    contract {
        callsInPlace(g)
    }
    return <!VIPER_VERIFICATION_ERROR!>f(g)<!>
}
```

The test case for that part of the code is the following one. The reported message signals that `g` may leak due calling `f`.

```
/calls_in_place_leak.kt:(663,667): warning: Function `g: () -> Unit` may leak from its contract.
```

Reply:
Ah, here it looks like it's actually correct.  The issue is in the other case:
```kotlin
fun leak(h: () -> Unit) = Unit
fun test(f: () -> Unit) {
  contract { callsInPlace(f) }
  leak(f)
}```
This will give a warning about `h` being leaked, instead of `f`.